Manatee.Json/Serialization/Internal/Serializers/JsonValueSerializer.cs

```diff
@@ -0,0 +1,154 @@
+using System;
+using System.Linq;
+using JetBrains.Annotations;
+
+namespace Manatee.Json.Serialization.Internal.Serializers
+{
+    [UsedImplicitly]
+    internal class JsonValueSerializer : IPrioritizedSerializer
+    {
+        public int Priority => 2;
+        public bool ShouldMaintainReferences => true;
+
+        private Type[] HandledTypes = {
+            typeof(JsonValue),
+            typeof(JsonObject),
+            typeof(JsonArray)
+        };
+
+        public bool Handles(SerializationContextBase context)
+        {
+            return HandledTypes.Contains(context.InferredType);
+        }
+
+        public JsonValue Serialize(SerializationContext context)
+        {
+            switch (context.Source)
```

Comment:
why are you switching on this and not just returning `context.Source`?  You need to return a `JsonValue` and you're getting a `JsonValue`...

Reply:
I understand it such that what comes in can be any of the three types the serializer can handle: `JsonValue`, `JsonArray` and `JsonObject`. For json literals, the JsonValue are returned, but for `JsonArray` and `JsonObject` I serialize each item/property. The reason for this is to maintain `"$ref"` functionality, just like the other serializers does.

I think if `"$ref"` is not important for `JsonValue`, `JsonArray` nor `JsonObject`, we could just return `context.Source` like you propose?

Reply:
If you're deserializing to the JSON content, object references are irrelevant.